apps/lib/chain_pusher/solana.go

```diff
@@ -0,0 +1,524 @@
+	slot, err := sci.client.GetSlot(context.Background(), rpc.CommitmentFinalized)
+	if err != nil {
+		return 0, err
+	}
+	//get the last 40 blocks to comply with rate limit
+	sci.limiter.Wait(context.Background())
+	blocks, err := sci.client.GetBlocks(context.Background(), slot-40, &slot, rpc.CommitmentFinalized)
+	if err != nil {
+		return 0, err
+	}
+
+	sci.limiter.Wait(context.Background())
+	blockTime, err := sci.client.GetBlockTime(context.Background(), blocks[0])
+	previousBlockTime := blockTime.Time()
+	if err != nil {
+		return 0, err
+	}
+	blocks = blocks[1:]
+	blockTimes := []time.Duration{}
+	for _, block := range blocks {
+		sci.limiter.Wait(context.Background())
+		currentBlock, err := sci.client.GetBlockTime(context.Background(), block)
+		if err != nil {
+			return 0, err
+		}
+		currentBlockTime := currentBlock.Time()
+		blockTimes = append(blockTimes, currentBlockTime.Sub(previousBlockTime))
+		previousBlockTime = currentBlockTime
+	}
+	var totalTime time.Duration
+	for _, blockTime := range blockTimes {
+		totalTime += blockTime
+	}
+
+	if len(blockTimes) == 0 {
+		return 0, fmt.Errorf("no block times available")
+	}
+
+	averageBlockTime := totalTime / time.Duration(len(blockTimes))
+	sci.logger.Debug().Dur("averageBlockTime", averageBlockTime).Msg("Average block time")
+	return averageBlockTime, nil
+}
+
+func (sci *SolanaContractInteracter) runUnboundedConfirmationBuffer(confirmationOutChan chan solana.Signature) {
+	var queue []solana.Signature
+	for {
+		if len(queue) == 0 {
+			sig := <-sci.confirmationInChan
+			queue = append(queue, sig)
+		}
+		select {
+		case sig := <-sci.confirmationInChan:
+			queue = append(queue, sig)
+		case confirmationOutChan <- queue[0]:
+			queue = queue[1:]
+		}
+	}
+}
+func (sci *SolanaContractInteracter) startConfirmationWorkers(ch chan solana.Signature, numWorkers int) {
+	for i := 0; i < numWorkers; i++ {
+		go sci.confirmationWorker(ch)
+	}
+	sci.logger.Info().Int("numWorkers", numWorkers).Msg("Started confirmation workers")
+}
+
+func (sci *SolanaContractInteracter) confirmationWorker(ch chan solana.Signature) {
+	for sig := range ch {
+		_, err := confirm.WaitForConfirmation(context.Background(), sci.wsClient, sig, nil)
+		if err != nil {
+			sci.logger.Error().Str("signature", sig.String()).Err(err).Msg("failed to confirm transaction")
+		} else {
+			sci.logger.Debug().Str("signature", sig.String()).Msg("confirmed transaction")
+		}
+	}
+}
+
+func (sci *SolanaContractInteracter) ListenContractEvents(ch chan map[InternalEncodedAssetId]InternalStorkStructsTemporalNumericValue) {
+	wg := sync.WaitGroup{}
+	for _, feedAccount := range sci.feedAccounts {
+		wg.Add(1)
+		go sci.listenSingleContractEvent(ch, feedAccount, &wg)
+	}
+	// Wait indefinitely
+	wg.Wait()
+}
+
+func (sci *SolanaContractInteracter) listenSingleContractEvent(ch chan map[InternalEncodedAssetId]InternalStorkStructsTemporalNumericValue, feedAccount solana.PublicKey, wg *sync.WaitGroup) {
```

Comment:
why do we need these `Internal` types? are they different from the existing non-internal types?

Reply:
These types were already present in the pusher and I patterned their use off of evm. My assumption is have a standard type that doesn't come from autogenerated bindings. The evm type seems to be the same but the solana type is slightly different. I'll defer to @akawalsky on this though.

Reply:
Confirming Henry's explanation is correct